core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -9,56 +11,192 @@ package scalaz.zio
 sealed abstract class ExitResult[+E, +A] extends Product with Serializable { self =>
   import ExitResult._
 
-  final def succeeded: Boolean = self match {
-    case Completed(_) => true
-    case _            => false
+  final def toEither: Either[Throwable, A] = self match {
+    case Completed(value) => Right(value)
+    case Failed(cause)    => Left(cause.toThrowable)
   }
 
-  final def map[B](f: A => B): ExitResult[E, B] = self match {
-    case Completed(a) => Completed(f(a))
-    case x            => x.asInstanceOf[ExitResult[E, B]]
+  final def leftMap[E1](f: E => E1): ExitResult[E1, A] =
+    self match {
+      case e @ Completed(_) => e
+      case Failed(c)        => Failed(c.map(f))
+    }
+
+  final def map[A1](f: A => A1): ExitResult[E, A1] =
+    self match {
+      case Completed(v)  => Completed(f(v))
+      case e @ Failed(_) => e
+    }
+
+  final def bimap[E1, A1](f: E => E1, g: A => A1): ExitResult[E1, A1] =
+    leftMap(f).map(g)
+
+  final def zip[E1 >: E, B](that: ExitResult[E1, B]): ExitResult[E1, (A, B)] = zipWith(that)((_, _), _ ++ _)
+
+  final def zipPar[E1 >: E, B](that: ExitResult[E1, B]): ExitResult[E1, (A, B)] = zipWith(that)((_, _), _ && _)
+
+  final def zipWith[E1 >: E, B, C](that: ExitResult[E1, B])(
+    f: (A, B) => C,
+    g: (Cause[E], Cause[E1]) => Cause[E1]
+  ): ExitResult[E1, C] =
+    (self, that) match {
+      case (Completed(a), Completed(b)) => Completed(f(a, b))
+      case (Failed(l), Failed(r))       => Failed(g(l, r))
+      case (e @ Failed(_), _)           => e
+      case (_, e @ Failed(_))           => e
+    }
+
+  final def exceptions: Option[List[Throwable]] = self match {
+    case Completed(_)  => None
+    case Failed(cause) => Some(cause.unchecked)
+  }
+
+  final def <>[E1, A1 >: A](that: ExitResult[E1, A1]): ExitResult[E1, A1] = self.failure match {
+    case Some(_) => that
+    case None    => self.asInstanceOf[ExitResult[E1, A1]]
   }
 
-  final def mapError[E2, A1 >: A](f: E => ExitResult[E2, A1]): ExitResult[E2, A1] = self match {
-    case ExitResult.Failed(e, _) => f(e)
-    case x                       => x.asInstanceOf[ExitResult[E2, A1]]
+  final def failure: Option[E] = self match {
```

Comment:
If you don't lose this, delete it, because it loses information. At the very least, change the name to `checked`.